### ISSUE TITLE
null/undefined are not acting as discriminators with object types

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#24479]
Found after discussing this DefinitelyTyped issue: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25731#issuecomment-391539084

```ts
export interface Errored {
    error: Error
    value: null
}

export interface Succeeded<Value> {
    error: null
    value: Value
}

type Result<T> = Succeeded<T> | Errored;

declare function getVal<T>(x :T): Result<T>

let x = getVal("hello");

if (x.error === null) {
//  vvvvvvv ERROR
    x.value.toUpperCase();
} 
```

Expected: No error
Actual: `x.value` is possibly `null`.

**Comment by DanielRosenwasser** [comment]
Duplicate of #24193.
